app/forms/firewall-rules-common.tsx

```diff
@@ -281,22 +284,22 @@ const ProtocolField = ({ control, protocol }: ProtocolFieldProps) => (
 
 type CommonFieldsProps = {
   control: Control<FirewallRuleValues>
-  project: string
-  instances: Array<Instance>
-  vpcs: Array<Vpc>
   nameTaken: (name: string) => boolean
   error: ApiError | null
 }
 
-export const CommonFields = ({
-  control,
-  project,
-  instances,
-  vpcs,
-  nameTaken,
-  error,
-}: CommonFieldsProps) => {
+export const CommonFields = ({ control, nameTaken, error }: CommonFieldsProps) => {
   const targetAndHostDefaultValues: TargetAndHostFormValues = { type: 'vpc', value: '' }
+  const { project } = useProjectSelector()
+  // prefetchedApiQueries below are prefetched in firewall-rules-create and -edit
+  const { data: instancesData } = usePrefetchedApiQuery('instanceList', {
+    query: { project, limit: PAGE_SIZE },
+  })
+  const instances = instancesData?.items ?? []
+  const { data: vpcData } = usePrefetchedApiQuery('vpcList', {
+    query: { project, limit: PAGE_SIZE },
+  })
+  const vpcs = vpcData?.items ?? []
```

Comment:
neither of these can be undefined because of the prefetch. you should be able to do

```ts
const { data: { items: vpcs } } = ...
```